juju/juju
# apiserver/uniter: Don't track container-scoped relations to applications not relevant for subordinate unit

## Description of change

If a subordinate application was related to more than one other app, units of the subordinate app wouldn't be cleaned up unless relations to *all* other apps were removed. This was because the uniter would track life for relations to all of the apps, even though the only relation that should keep this unit alive is the relation to the app that it's deployed alongside. For example, if a log monitoring app is related to mysql and mediawiki so that the units are as follows:
```
mysql/0
    logging/0
mediawiki/0
    logging/1 
```

... the logging/1 unit shouldn't track the life of the mysql-logging relation. If the mediawiki-logging relation dies then it should destroy itself.

This problem would also prevent you from removing one of the principal applications - destroying the application would hang waiting for the subordinate units to die. (This is the cause of bug 1655486.)

Change the uniter API so that a unit watches relations for a unit rather than at the application level, and make the `WatchUnitRelations` method filter out events for container-scoped relations to applications other than the unit's principal.

This change is added to the new v5 of the UniterAPI. There was some confusion in the versioning - the current version was registered as v4 and v5 (new methods would show up in the old version which causes problems for libjuju), there was a v3 in the code not linked to anything. This has been sorted out.

## QA steps

For the first bug:
* Deploy mysql, wordpress and nrpe to a model, and relate mysql and wordpress to nrpe. nrpe units will be created on each of the mysql and wordpress unit machines. 
* Remove the relation between mysql and nrpe.
* The mysql nrpe unit should be removed, but the wordpress one should remain.

To confirm the second bug is fixed:
* Deploy mysql, wordpress and ntp to a model. Make sure to specify `--series trusty` for mysql and ntp to ensure that everything can be linked up.
* Relate mysql to wordpress, ntp to mysql, ntp to wordpress.
* Once everything has settled down, remove wordpress - the wordpress unit and its subordinate ntp unit should be cleaned up and the application should be removed successfully.

## Bug reference

Fixes https://bugs.launchpad.net/juju/+bug/1686696 and https://bugs.launchpad.net/juju/+bug/1655486
